### ISSUE TITLE
Not running with readOnlyRootFilesystem set to true

### ISSUE THREAD

**Issue by enibache** [coreruleset/modsecurity-crs-docker#172]
I'm trying to run this image in my helm chart running in OCP4 whereby we have a quality gate that only allows pods with readOnlyRootFilesystem set to true. Is there perhaps hints on how to accomplish this using this image in my deployment.yaml?

**Comment by enibache** [comment]
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-generate-certificate.sh
/usr/local/bin/generate-certificate: using existing key / certificate
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: can not modify /etc/nginx/conf.d/default.conf (read-only file system?)
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/conf.d/default.conf.template to /etc/nginx/conf.d/default.conf
/docker-entrypoint.d/20-envsubst-on-templates.sh: line 33: can't create /etc/nginx/conf.d/default.conf: Read-only file system

**Comment by fzipi** [comment]
There are two problems: 
- how we create the container image, which for the nginx case needs to write files around.
- the logs must be written somewhere

The second one can be solved by using a persistent volume like:

**Comment by fzipi** [comment]
The [nginx ingress controller](https://docs.nginx.com/nginx-ingress-controller/configuration/security/) solves the first part by ... copying everything to a mounted directory:
```
#      volumes:
#      - name: nginx-etc
#        emptyDir: {}
#      - name: nginx-cache
#        emptyDir: {}
#      - name: nginx-lib
#        emptyDir: {}
#      - name: nginx-log
#        emptyDir: {}
.
.
.
#          readOnlyRootFilesystem: true
.
.
.
#        volumeMounts:
#        - mountPath: /etc/nginx
#          name: nginx-etc
#        - mountPath: /var/cache/nginx
#          name: nginx-cache
#        - mountPath: /var/lib/nginx
#          name: nginx-lib
#        - mountPath: /var/log/nginx
#          name: nginx-log
.
.
.
#      initContainers:
#      - image: <repository>:<tag>
#        imagePullPolicy: IfNotPresent
#        name: init-nginx-ingress
#        command: ['cp', '-vdR', '/etc/nginx/.', '/mnt/etc']
#        securityContext:
#          allowPrivilegeEscalation: false
#          readOnlyRootFilesystem: true
#          runAsUser: 101 #nginx
#          runAsNonRoot: true
#          capabilities:
#            drop:
#            - ALL
#        volumeMounts:
#        - mountPath: /mnt/etc
#          name: nginx-etc
```

**Comment by theseion** [comment]
#191 solves the root user issue.

**Comment by ne20002** [comment]
I just faced problems with the latest image version. I use my own nginx.conf mapped as nginx.conf.template.
After changing pid file location it worked again. 

For the readOnlyFilesystem I suggest to take a look at redis container, which I run with read-only mode.
Redis maps a tmpfs within the container which might be a solution for the crs image as well. If the created config would be in a tmpfs (its not much size) this may solve the read-only problem.

**Comment by ne20002** [comment]
Having the modsecurity-crs container able to run in read-only mode would be a great improvement.

**Comment by fzipi** [comment]
Latest pushed images should include the fix (E.g. ghcr.io/coreruleset/modsecurity-crs:3.3.5-nginx-202402070602), so please check and come back if something is not working.

**Comment by theseion** [comment]
@fzipi No, I don't think so. @ne20002 is talking about read-only filesystems, while the change is simply to use a non-root user (which has some similar implications, of course).

**Comment by fzipi** [comment]
You are right. Why I thought this was fixed? 🤔 Nevermind. Thanks for bringing this one back.

**Comment by fzipi** [comment]
Maybe we can add a similar chart with examples for k8s?

**Comment by theseion** [comment]
What kind of chart are you thinking about? A Helm chart? Or are you talking about the examples for the ingress controller you gave above?

**Comment by ne20002** [comment]
I tried to run the container with --read-only parameter (aka readOnlyRootFileSystem) and found:

- log files: log files can be easily written by mounting an external folder
- tmp and cache folders: can also be mounted or set as tmpfs
- the way the config files are created during startup. 
This seems to apply to folders `conf.d`, `includes` and `modsecurity.d` in `/etc/nginx/` and to the `nginx.conf` file itself.
Mounting the three folders as tmpfs fixed the errors but I found no way to fix the problems with `nginx.conf`.

In my setup I use my own nginx.conf which is passed as nginx.conf.template to the container. 

A simple option to let a user run the container with read-only file system would be to provide a parameter that disables the creation of the config files leaving the user responsible for mounting the configuration manually (which is not really a hard thing to do). At least, it would solve it for my use case.

A more general solution would be to move everything from `/etc/nginx` into another folder, mount `/etc/nginx` as tmpfs in the container/image by default and copy the whole configuration during startup as it is now (but just from a folder outside of `/etc/nginx`).

I believe, this should do the trick.

**Comment by theseion** [comment]
Thanks @ne20002. I've opened a PR: https://github.com/coreruleset/modsecurity-crs-docker/pull/210

**Comment by fzipi** [comment]
Can we close this one now?

**Comment by theseion** [comment]
No, there's still an error when running the image that I need to fix.

**Comment by mcpovel** [comment]
Just trying to run image as read only with those tmpfs mentioned for PR210 but failing already when trying to use tmpfs even without any of the read only mounst. Just using:
    tmpfs:
      - /tmp:rw,mode=777
      - /opt/owasp-crs:rw,mode=777 
      - /var/cache/nginx:rw,mode=777
      - /etc/nginx:rw,mode=777
      - /etc/modsecurity.d:rw,mode=777

Will already break 90-copy-modsecurity-config.sh

And even without any change for  /etc/modsecurity.d will fail on /etc/nginx beeing tmpfs due to:
grep: /etc/modsecurity.d/owasp-crs/crs-setup.conf: No such file or directory
Failed to find rule 900000 to set blocking_paranoia_level=1 for BLOCKING_PARANOIA in /etc/modsecurity.d/owasp-crs/crs-setup.conf. Aborting

**Comment by theseion** [comment]
Yes, there are still a couple of issues to fix. I'm working on it.

**Comment by domi27** [comment]
Would be also very interested in a solution for this setup :-)

**Comment by fzipi** [comment]
Something worth knowing while this is open: the read-only variants exist in the code but are not being built.

`docker-bake.hcl` carries the matrix entry commented out:

```hcl
    read-only-fs = [
        {
            name = "writable"
            read-only = "false"
        },
        # {
        #     name = "read-only"
        #     read-only = "true"
        # }
    ]
```

Meanwhile the README documents `nginx-read-only` and `nginx-alpine-read-only` under "Read-only Root Filesystem" as if they were published, and `READ_ONLY_FS` plus `nginx/docker-entrypoint.d/0-move-writables.sh` are all still in place. So anyone following the README looking for those tags will not find them.

Either the variants get built again or the README section should say they are unavailable. Which way do you want it, @fzipi? Re-enabling the matrix entry is a small change, but it triples the nginx build matrix, and the entrypoint logic would need verifying against the current image layout before publishing tags people rely on.

For the OpenShift case specifically, @dchandekstark's recipe above (chgrp 0 plus chmod g=u on the writable paths) is the standard approach and does not need a read-only image.

**Comment by theseion** [comment]
We never finished the readonly images so that the reliably worked. So rather stop advertising in the docs.
